### PR TITLE
V2 X3D Serializer

### PR DESCRIPTION
**Pull request by z3dev**
This pull request updates the X3D serializer to use CSG V2 API. The goals of this pull request are:
- [X] Update to use CSG V2 API
- [X] Update code up to JSCAD coding standards
- [X] Add test suites to complete coverage, 98.72%

### All Submissions:

* [X] Have you followed the guidelines in our Contributing document?
* [X] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?
* [X] Does your submission pass tests?

### CODE REVIEW

File: packages/io/x3d-serializer/README.md [hunk 1]
```diff
@@ -1,15 +1,17 @@
...
 
 ## Overview
 
-This serializer outputs a 'blobable' array of data (from a CSG object)
-ie an array that can either be passed directly to a Blob (`new Blob(blobable)`)
-or converted to a Node.js buffer.
+This serializer outputs a 'blobable' array of data from one or more JSCAD geometries.
+The array of data can either be used to create a Blob (`new Blob(blobable)`), or converted to a Node.js buffer.
+
+The serialization of the following geometries are possible.
```
Comment (kaosat-dev) [inline packages/io/x3d-serializer/README.md]: good !

File: packages/io/x3d-serializer/index.js [hunk 2]
```diff
@@ -33,102 +32,119 @@ const stringify = require('onml/lib/stringify')
...
   options = Object.assign({}, defaults, options)
 
-  options.statusCallback && options.statusCallback({progress: 0})
+  objects = utils.flatten(objects)
+
+  // convert only 3D geometries
+  let objects3d = objects.filter((object) => geometry.geom3.isA(object))
+
+  if (objects3d.length === 0) throw new Error('only 3D geometries can be serialized to X3D')
+  if (objects.length !== objects3d.length) console.warn('some objects could not be serialized to X3D')
```
Comment (kaosat-dev) [inline packages/io/x3d-serializer/index.js]: good !

File: packages/io/x3d-serializer/index.js [hunk 3]
```diff
@@ -33,102 +32,119 @@ const stringify = require('onml/lib/stringify')
...
+      }
     }
   })
   scene = scene.concat(shapes)
   return [scene]
 }
 
-const convertCSG = (object, options) => {
-  var shape = ['Shape', {}, convertMesh(object, options)]
+const convertShape = (object, options) => {
```
Comment (kaosat-dev) [inline packages/io/x3d-serializer/index.js]: smart naming :)

File: packages/io/x3d-serializer/index.js [hunk 4]
```diff
@@ -33,102 +32,119 @@ const stringify = require('onml/lib/stringify')
...
-    ['Coordinate', {point: pointList}],
-    ['Color', {color: colorList}]
+    { ccw: 'true', colorPerVertex: 'false', solid: 'false', index: indexList },
+    ['Coordinate', { point: pointList }],
+    ['Color', { color: colorList }]
   ]
   return faceset
 }
 
+const convertToTriangles = (object, options) => {
```
Comment (kaosat-dev) [inline packages/io/x3d-serializer/index.js]: nice !
But I am a bit worried about this one: do we not have a more generic need for triangulated meshes ? (that is needed from webgl viewing , most 3d exports etc), so put another way: should we not have a geom3.toTriangles ? (that can use this code for example ?)

File: packages/io/x3d-serializer/tests/serialize.test.js [hunk 5]
```diff
@@ -57,9 +59,9 @@ const expected3 = `<?xml version="1.0" encoding="UTF-8"?>
       </IndexedTriangleSet>
     </Shape>
     <Shape>
-      <IndexedTriangleSet ccw=\"true\" colorPerVertex=\"false\" solid=\"false\" index=\"0 1 2 0 3 1 4 5 6 4 7 5 0 6 3 0 4 6 2 5 7 2 1 5 0 7 4 0 2 7 3 5 1 3 6 5\">
-        <Coordinate point=\"4 4 4 4 6 6 4 6 4 4 4 6 6 4 4 6 6 6 6 4 6 6 6 4\"/>
-        <Color color=\"0.5 1 0.5 0.5 1 0.5 0.5 1 0.5 0.5 1 0.5 0.5 1 0.5 0.5 1 0.5 0.5 1 0.5 0.5 1 0.5 0.5 1 0.5 0.5 1 0.5 0.5 1 0.5 0.5 1 0.5\"/>
+      <IndexedTriangleSet ccw="true" colorPerVertex="false" solid="false" index="0 1 2 0 3 1 4 5 6 4 7 5 0 6 3 0 4 6 2 5 7 2 1 5 0 7 4 0 2 7 3 5 1 3 6 5">
```
Comment (kaosat-dev) [inline packages/io/x3d-serializer/tests/serialize.test.js]: the newer version seems more logical and does not have the weird backlashes, but is it correct X3D ?

Comment (z3dev) [reply packages/io/x3d-serializer/tests/serialize.test.js]: Ahh... These changes from from eslint, which suggested to remove those backslashes.

Correct X3D? If there's an official 'validator' there we could make sure the serialized designs are correct X3D.

File: packages/io/x3d-serializer/index.js (hunk 4 above)
Comment (z3dev) [reply packages/io/x3d-serializer/index.js]: I think that a toTriangles() function would be a better solution. As we move to a 'descriptive' model then direct access to the polygons should not be possible. Conversion of the 'description' model to data should be requested via a function call, i.e. toPolygons(), toTriangles(), etc.

Not sure if we should do this now. SIGH

Comment (kaosat-dev) [reply packages/io/x3d-serializer/index.js]: Agreed but It is not urgent, I will add it to my "V2 changes and tweaks" doc for now :-)

File: packages/io/x3d-serializer/tests/serialize.test.js (hunk 5 above)
Comment (kaosat-dev) [reply packages/io/x3d-serializer/tests/serialize.test.js]: No clue, I barely ever heard of x3d at all before the serializer to be honest :-) 
Seems like cleaner XML in the V2 version though so all good :-)

**Pull request merged**